plat/common/lcpu.c

```diff
@@ -48,3 +165,424 @@ void ukplat_lcpu_halt_to(__nsec until)
+{
+	__lcpuidx this_cpu_idx = (__lcpuidx) ukplat_lcpu_id();
+
+	UK_ASSERT(this_cpu_idx < lcpu_count);
+	UK_ASSERT(lcpus[this_cpu_idx].idx == this_cpu_idx);
+	return this_cpu_idx;
+}
+#else /* CONFIG_UKPLAT_LCPU_IDISIDX */
+__lcpuidx ukplat_lcpu_idx(void)
+{
+	__lcpuid this_cpu_id  = ukplat_lcpu_id();
+	__u32 i, this_cpu_idx = (__u32) (-1);
+
+	/** TODO: Until we have a better way just do a linear search */
+	for (i = 0; i < lcpu_count; i++)
+		if (lcpus[i].id == this_cpu_id) {
+			UK_ASSERT(lcpus[i].idx == (__lcpuidx) i);
+			this_cpu_idx = i;
+			break;
+		}
+
+	UK_ASSERT(this_cpu_idx != (__u32) (-1));
+
+	return (__lcpuidx) this_cpu_idx;
+}
+#endif /* CONFIG_UKPLAT_LCPU_IDISIDX */
+
+int lcpu_fn_enqueue(struct lcpu *lcpu, const struct ukplat_lcpu_func *fn)
+{
+	void (*old_fn)(struct __regs *, const struct ukplat_lcpu_func *);
+
+	UK_ASSERT(fn->fn);
+
+	old_fn = ukarch_load_n(&lcpu->fn.fn);
+
+	/* Check if the slot is empty */
+	if (old_fn != NULL)
+		return -EAGAIN;
+
+	/* It is empty, try to store the function */
+	if (ukarch_compare_exchange_sync(&lcpu->fn.fn, old_fn,
+					 fn->fn) != fn->fn)
+		return -EAGAIN;
+
+	/* We have acquired the slot! Also store the user argument.
+	 * It is safe to do it afterwards, because the RUN IRQ handler will
+	 * only take one function and return afterwards. And we only raise the
+	 * IRQ after finishing setup.
+	 */
+	lcpu->fn.user = fn->user;
+
+	/* Ensure everything is written back when we return and the arch
+	 * support code will raise the IRQ
+	 */
+	wmb();
+
+	return 0;
+}
+
+static void lcpu_fn_dequeue(struct lcpu *this_lcpu, struct ukplat_lcpu_func *fn)
+{
+	*fn = this_lcpu->fn;
+
+	/* Ensure that we have captured the whole function object */
+	rmb();
+
+	UK_ASSERT(fn->fn);
+
+	/* Free the slot. Another function object can be queued afterwards */
+	this_lcpu->fn.fn = NULL;
+}
+
+static int lcpu_ipi_run_handler(void *args __unused, struct __regs *regs)
+{
+	struct lcpu *this_lcpu = lcpu_get_current();
+	struct ukplat_lcpu_func fn;
+
+	lcpu_fn_dequeue(this_lcpu, &fn);
+	fn.fn(regs, &fn);
+
+	/* If we had a transition from BUSY to HALTED in fn, we would
+	 * not reach this code but sit in the error halt loop. We can
+	 * thus safely just decrement without worrying about the HALTED
+	 * state.
+	 */
+	UK_ASSERT(lcpu_state_is_busy(this_lcpu->state));
+	ukarch_dec(&this_lcpu->state);
+
+	return 1;
+}
+
+static int lcpu_ipi_wakeup_handler(void *args __unused,
+				   struct __regs *regs __unused)
+{
+	/* Nothing to do */
+	return 1;
+}
+
+int lcpu_mp_init(void *arg)
+{
+	int rc;
+
+	/* Make sure this is run on the BSP only */
+	UK_ASSERT(lcpu_count == 1);
+	UK_ASSERT(lcpu_get_current() == lcpu_get(0));
+
+	/* Initialize architecture-dependent functionality. This will also do
+	 * CPU discovery and allocation
+	 */
+	rc = lcpu_arch_mp_init(arg);
+	if (unlikely(rc))
+		return rc;
+
+	/* Register the lcpu_run and lcpu_wakeup interrupt handlers */
+	rc = ukplat_irq_register(LCPU_ARCH_IPI_RUN_IRQ,
```

Comment:
So we still call use the term `ARCH` in the name of the IRQ definitions? I really think they should not be architecture-defined, but platform defined. Perhaps change the naming to `PLAT` and add a check in the platform `lcpu.h`as:
```
#ifdef CONFIG_HAS_SMP
#if !defined(LCPU_PLAT_IPI_RUN_IRQ) || !defined(LCPU_PLAT_IRI_WAKEUP_IRQ)
#error LCPU_PLAT_IPI_RUN_IRQ and LCPU_PLAT_IPI_WAKEUP_IRQ are not defined by the platform
#endif
#endif /* CONFIG_HAS_SMP */
```
The platform still has the choice to hardcode or provide KConfig options.

Reply:
Good point and we should change it. But it would not work like you suggest. At the moment the common code is in the `common` directory and imports the _common_ architecture-specific header. A certain platform is not mentioned anywhere. Instead, the platform imports the common API and defines the symbols. So what we can do is define two variables `lcpu_plat_ipi_run/wakeup_irq`. That are then referenced here. The platform could initialize them using a config option or even use some boot code to dynamically find the right IRQ numbers.

Reply:
Works for me 👍🏼

Reply:
Those interrupts aren't really dependent on the platform. They should be the same for every platform of the same architecture, in my opinion, so it would make sense to be defined at architecture level

Reply:
Yes, right. I also think that we will probably have the same IRQ vectors for all platforms of a certain architecture. But it is fair to keep it flexible. To not introduce further symbols, I simply made it parameters to `lcpu_mp_init`. Then you can compute the IRQ values at runtime or just pass `CONFIG_` defines.